### DISCUSSION TITLE
Support broticffi in addition to brotli

### DISCUSSION THREAD

**Discussion by mgorny** [aio-libs/aiohttp#7610] [Ideas] [answered]
aiohttp currently supports only the `brotli` bindings for Brotli. Unfortunately, these [stopped working on PyPy recently](https://github.com/google/brotli/issues/1072) and, well, we haven't been really able to figure out what's wrong. So far the only suggestion was to use `brotlicffi` instead (which might be better suited for PyPy anyway).

Would you be interested in adding support for `brotlicffi` alternatively to `brotli`? I can make a pull request. Just let me know whether you prefer the "fallback" approach (i.e. trying `brotli`, then `brotlicffi` on `ImportError`), or switching between `brotli` and `brotlicffi` depending on whether CPython or PyPy is being used.

TIA.

**Comment by Dreamsorcerer** [comment, marked as answer]
I'm not sure how it's implemented currently, so not sure what approach is possible.
But, I think best approach would be to have an easy method to plug in any custom library, if that's possible (as we've just done with charset guessing in the upcoming 3.9).
If that's not possible, then probably what is recommended on their readme: https://pypi.org/project/brotlicffi/#using-brotlicffi-in-projects
(I started to write my thoughts down, and then looked at their readme, and it recommended exactly what I was thinking).

**Reply by mgorny** [reply]
Thanks. I'm going to try the latter approach, and try to do minimal changes to the existing code.
